### PR TITLE
feat: feed messages to state from MQ

### LINKED ISSUE #66: feat: feed MQ messages to rollup-state-manager, and then write to prover-cluster DB
## STEP 1

change rollup-state-manager/tests/global_state/gen_testcase.rs#L74 `load_msgs_from_file` to `load_msgs_from_mq`
(we would still like to keep rollup-state-manager/tests/global_state/gen_testcase.rs there , so it'd be better to implement a new rollup-state-manager/src/bin/main.rs following the similar logic in rollup-state-manager/tests/global_state/gen_testcase.rs

Requirements:

1. implement `load_msgs_from_mq`

**keep `load_msgs_from_mq` running** (like, forever :P), read messages from MQ.
The MQ is set up by dingir-exchange, and its endpoint is https://github.com/Fluidex/dingir-exchange/blob/master/config.yaml#L4 . You may borrow some codes from https://github.com/Fluidex/dingir-exchange/blob/master/src/bin/test_unifymessenger.rs to read messages.

which messages to read? (let's work on these 3 first, and leave others as future work)
+ types::matchengine::messages::BalanceMessage (topic: "orders")
+ types::matchengine::messages::TradeMessage (topic: "trades")
+ types::matchengine::messages::OrderMessage (topic: "orders")

2. when receiving messages, use `_sender: crossbeam_channel::Sender<WrappedMessage>` to send `WrappedMessage` to `msg_receiver` via channel.

the messages then will be processed by `replay_msgs`

`replay_msgs` converts `WrappedMessage` to `L2Block` and then `L2Block` will be sent to `block_receiver` using `block_sender`

We would also need to modify `replay_msgs` a little bit.

Previously, `msg_receiver.iter()` collects all the messages parsed from `load_msgs_from_file`, and then run `witgen.flush_with_nop();` at the end.

We now would like to **keep `replay_msgs` running**, and then `witgen.flush_with_nop()` every `*test_utils::params::NTXS`.

block_receiver will recieve some data to write into DB.

(So in this sense we write into DB every time we get a `L2Block` ready, that is, every time we receive `*test_utils::params::NTXS` messages)

+ which DB to write into?
    * https://github.com/Fluidex/prover-cluster/blob/master/docker/docker-compose.yaml
+ how to write into the DB? in which format?
    * there is only 1 table in the DB, defined here: https://github.com/Fluidex/prover-cluster/blob/master/migrations/20210330142700_task.sql
    * we already have `L2Block`s generated by witgen, follow https://github.com/Fluidex/rollup-state-manager/blob/master/tests/global_state/export_circuit.rs#L34 to retrieve `input` and `ouput`, we need to write each  `input`+`ouput` pair into DB
        - for now you could just use unique task_id，I will adding auto_increment stuff in prover-cluster DB later
        - https://github.com/Fluidex/prover-cluster/blob/832f688759eef9b0c4454ba6e3061425eaed7a5f/migrations/20210330142700_task.sql#L5 is `block` for now
        - for the insert syntax you could refer to: https://github.com/Fluidex/prover-cluster/blob/832f688759eef9b0c4454ba6e3061425eaed7a5f/migrations/20210330142700_task.sql#L36

### PR DESCRIPTION
**Pull request by silathdiir**
Step 1 of #66 

### Summary

1. Adds file `tests/global_state/main_loop.rs`.
2. Copies `msg_consumer.rs` from https://github.com/Fluidex/dingir-exchange/blob/master/src/message/consumer.rs.
3. Implements function `load_msgs_from_mq` in `msg_loader.rs`.
4. Implements function `MessageWriter::on_message` to send received messages to function `replay_msgs`.
5. Implements function `save_block_to_db`.

### CODE REVIEW

**Comment by lispc** [comment]
@silathdiir add a line "// copied from https://github.com/Fluidex/dingir-exchange/blob/master/src/message/consumer.rs" as the first line in the copied src file

File: tests/global_state/main_loop.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            }
+
+            let new_block_num = witgen.get_block_generate_num();
+            if new_block_num > current_block_num {
+                current_block_num = new_block_num;
+                println!(
+                    "genesis {} blocks (TPS: {})",
+                    current_block_num,
+                    (*params::NTXS * current_block_num) as f32 / timing.elapsed().as_secs_f32()
+                );
```
Comment (lispc) [inline tests/global_state/main_loop.rs:73]: change log to println!("generate {} blocks with block_size {} in {}s: average TPS: {}", current_block_num, *params::NTXS, timing.elapsed().as_secs_f32() here

File: tests/global_state/msg_loader.rs [hunk 2]
```diff
@@ -18,7 +22,80 @@ pub fn load_msgs_from_file(
...
+                        log::info!("Ctrl-c received, shutting down");
+                        break;
+                    },
+
+                    err = cr_main.run_stream(|cr|cr.stream()) => {
+                        log::error!("Kafka consumer error: {}", err);
+                    }
+                }
+            }
+        });
```
Comment (HAOYUatHZ) [inline tests/global_state/msg_loader.rs:69]: this seems can be simplified?

Comment (HAOYUatHZ) [reply tests/global_state/msg_loader.rs:69]: But this is really not a big deal. We can refator this in the future.

**Review by HAOYUatHZ — APPROVED** [review]

**Pull request merged**